charms/worker/k8s/src/charm.py

```diff
@@ -515,10 +515,12 @@ def _get_snap_version(self) -> Optional[str]:
     )
     def _join_cluster(self):
         """Retrieve the join token from secret databag and join the cluster."""
-        if self.is_control_plane and self.api_manager.is_cluster_bootstrapped():
-            return
-
         if not (relation := self.model.get_relation("cluster")):
+            status.add(ops.BlockedStatus("Missing cluster integration"))
+            assert False, "Missing cluster integration"  # nosec
+
+        if self.is_control_plane and self.api_manager.is_cluster_bootstrapped():
+            relation.data[self.unit]["joined"] = self.get_node_name()
```

Comment:
The switch in the order is here b/c `cluster` as a peer relation will ALWAYS exist on control-plane units -- so that's a no-op and if it's not there on the worker -- let's just GIVE Up completely on the reconciliation loop and wait

Next -- this is important.  If i'm a control-plane unit and i think im in a bootstrapped cluster, lets just say i'm joined.  There's no harm here.  it may write the same relation data again -- but if the previous hook joined -- but we got some kind of 500 error from the snap like a context timeout -- then we REALLY did get joined, but just didn't finish the context method in `self.collector.joined(...)`  If we don't tell the leader control-plane that we are joined -- this  node will have a hard time leaving the cluster in the future